internal/engine/compiler/impl_amd64.go

```diff
@@ -3598,21 +3598,95 @@ func (c *amd64Compiler) compileLoadDataInstanceAddress(dataIndex uint32, dst asm
 	)
 }
 
-// compileMemoryCopy implements compiler.compileMemoryCopy for the amd64 architecture.
-func (c *amd64Compiler) compileMemoryCopy() error {
-	return c.compileCopyImpl(false, 0, 0)
+// compileCopyLoopImpl implements a REP MOVSQ memory copy for the given range with support for both directions.
+func (c *amd64Compiler) compileCopyLoopImpl(destinationOffset, sourceOffset, copySize *runtimeValueLocation, backwards bool, bwOffset uint8) {
+	// skip if nothing to copy
+	c.assembler.CompileRegisterToConst(amd64.CMPQ, copySize.register, 0)
+	emptyEightGroupsJump := c.assembler.CompileJump(amd64.JEQ)
+
+	rvalues := []*runtimeValueLocation{destinationOffset, sourceOffset, copySize}
+
+	// Prepare registers for swaps. There will never be more than 3 XCHGs in total.
+	swaps := c.compilePreventCrossedTargetRegisters(rvalues, []asm.Register{amd64.RegDI, amd64.RegSI, amd64.RegCX})
```

Comment:
As suggested, let's return the restore func, and then defer it so that we don't forget to invoke later on (as that requires the context later vs just defer here.

```suggestion
	restore := c.compilePreventCrossedTargetRegisters(rvalues, []asm.Register{amd64.RegDI, amd64.RegSI, amd64.RegCX})
	defer restore()
```